### PR TITLE
compile time lengths support in data stores

### PR DESCRIPTION
**Pull request by anstaf**
The changes is the storage API:
- Data store builder `dimensions` method now accepts both integers and integral constants. Type information is propagated to the data store that is built.
- The properly typed lengths and strides are available to the user via the `native_lengths()` and `native_strides()` methods from data_store/view/info.
- Data store sid adapter uses `native_lengths()`/`native_strides()` now. So the knowledge of the compile time lengths is propagated to the SID API
- call operator of the data store view is now also `compile time aware`. I.e. if the dimension offsets are provided as integral constants and all strides are also known in compile time, no stride computations are performed in run time.  
- There are minor backward incompatibilities to the public API:
  - `data_store` and `info` template classes are moved from the public `gridtools::storage` namespace. They should be considered implementation details now.
  - `lengths()` and `strides()` methods now return by value, not by const reference.

### CODE REVIEW

**Comment by anstaf** [comment]
launch tests

**Comment by anstaf** [comment]
@havogt Jenkins tests are OK, but cuda related cmake checks are failing and failed to diagnose why. Hannes, I probably need your expertise to figure out the root of the problem.

**Comment by havogt** [comment]
The failing tests are a problem in the docker containers, not a problem with your code. I need to spend more time debugging the issue that I introduced there...

File: include/gridtools/storage/info.hpp [hunk 1]
```diff
@@ -18,99 +18,136 @@
...
+            template <class Layout, class Align>
+            struct make_padded_length_f {
+                Align m_align;
+                template <class Length>
+                integral_constant<int, 1> operator()(integral_constant<int, -1>, Length) const {
+                    return {};
+                }
+                template <int Max = Layout::max_arg, class Length, std::enable_if_t<Max != 0 && Max != -1, int> = 0>
+                auto operator()(integral_constant<int, Layout::max_arg>, Length length) const {
+                    return (length + m_align - integral_constant<int, 1>()) / m_align * m_align;
```
Comment (havogt) [inline include/gridtools/storage/info.hpp:47]: Nice, I completely forgot that you implemented integral constant arithmetic.

File: tests/unit_tests/storage/test_storage_sid.cpp [hunk 2]
```diff
@@ -101,5 +102,16 @@ namespace gridtools {
...
+
+        namespace compile_time_lengths {
+            using testee_t =
+                decltype(storage::builder<storage_traits_t>.type<int>().layout<0, 1>().dimensions(10_c, 10_c)());
+
+            static_assert(sid::concept_impl_::is_sid<testee_t>(), "");
+
+            static_assert(tuple_util::is_empty_or_tuple_of_empties<sid::strides_type<testee_t>>(), "");
+            static_assert(tuple_util::is_empty_or_tuple_of_empties<sid::lower_bounds_type<testee_t>>(), "");
+            static_assert(tuple_util::is_empty_or_tuple_of_empties<sid::upper_bounds_type<testee_t>>(), "");
```
Comment (havogt) [inline tests/unit_tests/storage/test_storage_sid.cpp:114]: For completeness we could add checks that lengths and bounds have the correct static sizes.

**Review by havogt — CHANGES_REQUESTED** [review]
Only some minor comments.

File: include/gridtools/storage/builder.hpp [hunk 3]
```diff
@@ -34,12 +37,29 @@ namespace gridtools {
                 struct layout {};
             } // namespace param
 
-            template <class Fun, class T, class Layout, size_t N, size_t... Is>
-            void initializer_impl(
-                Fun const &fun, T *dst, Layout layout, info<N> const &info, std::index_sequence<Is...>) {
+            template <class, class = void>
+            struct is_integral_constant : std::false_type {};
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/storage/builder.hpp]: done

File: tests/unit_tests/storage/test_storage_sid.cpp (hunk 2 above)
Comment (anstaf) [reply tests/unit_tests/storage/test_storage_sid.cpp]: done

**Comment by anstaf** [comment]
launch jenkins

**Review by havogt — APPROVED** [review]

**Comment by havogt** [comment]
You can merge once jenkins is green. I am still working on the container image...

**Pull request merged**